autosklearn/ensemble_builder.py

```diff
@@ -671,12 +679,32 @@ def predict(self, set_: str,
         # TODO: ADD saving of predictions on "ensemble data"
 
     def _delete_non_candidate_models(self, candidates):
-        cand_ids = [cand.replace('_', '.').split('.')[-2] for cand in candidates]
-        for model_file in self.read_preds.keys():
-            model_file_id = model_file.split('.')[-2]
-            if model_file_id not in cand_ids:
-                os.remove(model_file)
-                self.logger.info("Deleted file of non-candidate model %s", model_file)
+        for pred_file in self.read_preds.keys():
+            if pred_file not in candidates and pred_file not in self.deleted_preds:
+                # Delete prediction file
+                try:
+                    pred_lock = pred_file + '.lock'
+                    with lockfile.LockFile(pred_lock):
```

Comment:
I added the locks, but I'm not quite sure that they are necessary. 
Before adding the locks, I ran the test_delete_non_candidate_models with very loose time constraints and got no errors.

Reply:
I'm not sure either, but let's better be safe than sorry.